### ISSUE TITLE
Visitor pattern for LK proofs

### ISSUE THREAD

**Issue by loewenheim** [gapt/gapt#431]
We've been talking about maybe implementing a visitor class for LK proofs. I've written a bit of code:
```scala
class lkVisitor {

  def apply(proof: LKProof): LKProof = withOccConnector(proof)._1

  def withOccConnector(proof: LKProof): (LKProof, OccConnector[HOLFormula]) = apply_(proof)

  protected def apply_(proof: LKProof): (LKProof, OccConnector[HOLFormula]) = proof match {
    case p @ InitialSequent(_) =>
      visitInitialSequent(p)

    case p @ WeakeningLeftRule( subProof, formula ) =>
      visitWeakeningLeft(p)

    case p @ WeakeningRightRule( subProof, formula ) =>
      visitWeakeningRight(p)

    case p @ ContractionLeftRule( subProof, formula ) =>
      visitContractionLeft(p)

    case p @ ContractionRightRule( subProof, formula ) =>
      visitContractionRight(p)

    case p @ CutRule( leftSubProof, aux1, rightSubProof, aux2 ) =>
      visitCut(p)
  }

  protected def visitInitialSequent(proof: InitialSequent): (LKProof, OccConnector[HOLFormula]) = (proof, OccConnector(proof.endSequent))

  protected def visitWeakeningLeft(proof: WeakeningLeftRule): (LKProof, OccConnector[HOLFormula]) = {
    val (subProofNew, subConnector) = apply_(proof.subProof)
    val proofNew = WeakeningLeftRule(subProofNew, proof.mainFormula)
    val connector = (proofNew.getOccConnector * subConnector * proof.getOccConnector.inv) + (proofNew.mainIndices(0), proof.mainIndices(0))

    (proofNew, connector)
  }

  protected def visitWeakeningRight(proof: WeakeningRightRule): (LKProof, OccConnector[HOLFormula]) = {
    val (subProofNew, subConnector) = apply_(proof.subProof)
    val proofNew = WeakeningRightRule(subProofNew, proof.mainFormula)
    val connector = (proofNew.getOccConnector * subConnector * proof.getOccConnector.inv) + (proofNew.mainIndices(0), proof.mainIndices(0))

    (proofNew, connector)
  }

  protected def visitContractionLeft(proof: ContractionLeftRule): (LKProof, OccConnector[HOLFormula]) = {
    val (subProofNew, subConnector) = apply_(proof.subProof)
    val List(aux1, aux2) = proof.auxIndices(0) map subConnector.child
    val proofNew = ContractionLeftRule(subProofNew, aux1, aux2)
    val connector = proofNew.getOccConnector * subConnector * proof.getOccConnector.inv

    (proofNew, connector)
  }

  protected def visitContractionRight(proof: ContractionRightRule): (LKProof, OccConnector[HOLFormula]) = {
    val (subProofNew, subConnector) = apply_(proof.subProof)
    val List(aux1, aux2) = proof.auxIndices(0) map subConnector.child
    val proofNew = ContractionRightRule(subProofNew, aux1, aux2)
    val connector = proofNew.getOccConnector * subConnector * proof.getOccConnector.inv

    (proofNew, connector)
  }

  protected def visitCut(proof: CutRule): (LKProof, OccConnector[HOLFormula]) = {
    val (leftSubProofNew, leftSubConnector) = apply_(proof.leftSubProof)
    val (rightSubProofNew, rightSubConnector) = apply_(proof.rightSubProof)
    val (aux1, aux2) = (leftSubConnector.child(proof.aux1), rightSubConnector.child(proof.aux2))
    val proofNew = CutRule(leftSubProofNew, aux1, rightSubProofNew, aux2)
    val connector = (proofNew.getLeftOccConnector * leftSubConnector * proof.getLeftOccConnector.inv) + (proofNew.getRightOccConnector * rightSubConnector * proof.getRightOccConnector.inv)

    (proofNew, connector)
  }

}
```

… and so on and so forth. What do you think, @gebner? Is this the way to go?

**Comment by gebner** [comment]
Hmm, you know that I don't like matching on InitialSequents. :smile: 

More seriously, this looks great for functions of type `LKProof => LKProof`; I like that you pass the actual proofs to the `visit*` methods (instead of the subproofs, auxiliary formulas, etc.), this way you can easily access other fields on the proofs.

AFAICT this visitor would simplify the following functions:
* substitution
* term replacement
* `cleanStructuralRules`
* definition elimination
* reductive cut-elimination
* regularization
* the substitution part in the new tactics prover

It would not work for:
* `moveStrongQuantifiersDown`
* Skolemization

The last two are functions `(LKProof, Sequent[A]) => LKProof`, do you have any ideas for them?

Thinking of `cleanStructuralRules`, what should the visitor do if an auxiliary formula is missing?

**Comment by gebner** [comment]
Further stuff to think about:

Should we include memoization to deal with daglike proofs?  (Admittedly this is more relevant for resolution, but it might be a good idea to have a similar visitor there, and memoization directly affects the design choice for the lifecycle of the visitor object.)

This is a different approach than visitors, but it might work nicely for a larger class of algorithms:  In Scala, case classes automatically have a copy method, e.g. if you have a `case class A(b: B, c: C)` then you can do `a.copy(b = anotherB)` to update the value of `b`.  For proofs, we could add a similar method: `p.copy(auxIndices = ..., immediateSubProofs = ...)`.  This approach would work for `moveStrongQuantifiersDown` and Skolemization as well.  Basically you would just have one default case in each algorithm which uses `.copy()` instead of very similar cases for all the propositional rules.  For further ease-of-use we could add methods like `.applyToSubProofs()` so that the default case would look like `case p => p.applyToSubProofs(apply)` and would automatically take care of the OccConnectors.  This also generalizes neatly to `.applyToSubProofsAndCopyLabels()`, `.applyToSubProofsAndWeakenMissingFormulas()`, etc.

**Comment by quicquid** [comment]
In OCaml the visitor pattern is implemented with a variable of parametrized type. The [Real World OCaml] (https://realworldocaml.org/v1/en/html/classes.html#open-recursion) book has some code examples. In our case I'd propose to add this return parameter also to the return value, s.t. the type of a function would be visit(p : CutRule, accumulator : T) : (LKProof, T). E.g. a memoization visitor would just add the memo table to the accumulator.

**Comment by loewenheim** [comment]
I've now implemented this in 9ca26bacf97c88c303f4311149069bf05e0c8bca. I think that functions of types other than `LKProof => LKProof`—i.e. those that need more arguments—aren't a problem if you derive a class with constructor arguments from the visitor trait, as opposed to an object. See the `eliminateTheoryAxioms` algorithm for an example.
